### PR TITLE
Boundary conditions 1 d

### PR DESCRIPTION
**Pull request by chillenzer**
Implement twisted boundary conditions for the 1D derivative operator.

### CODE REVIEW

**Comment by Mayank447** [comment]
I will test this code with example script on sine derivative defined over [0, pi). If it works will merge

**Comment by chillenzer** [comment]
Did it myself and, in fact, found a wrong normalisation. Now everything looks good.

File: test/test_boundary_conditions.py [hunk 1]
```diff
@@ -21,7 +21,7 @@ def spectrum(request):
 
 def test_eigenfunctions_obey_boundary_conditions(spectrum):
     eigenfunctions = spectrum.eigenfunction(np.arange(spectrum.total_num_lattice_points).reshape(-1, 1))
-    assert np.allclose(np.exp(2.0j * np.pi * spectrum.theta * spectrum.a) * eigenfunctions(*spectrum.lattice()), eigenfunctions(spectrum.lattice()[0] + spectrum.L))
+    assert np.allclose(np.exp(2.0j * np.pi * spectrum.theta) * eigenfunctions(*spectrum.lattice()), eigenfunctions(spectrum.lattice()[0] + spectrum.L))
```
Comment (Mayank447) [inline test/test_boundary_conditions.py:24]:
```suggestion
spectral_representation = np.eye(spectrum.total_num_lattice_points)
real_space_representation = spectrum.transform(spectral_representation, "spectral", "real")
eigenfunction = spectrum.eigenfunction(np.arange(spectrum.total_num_lattice_points))
```

File: test/test_boundary_conditions.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+def test_inverse_transform_results_in_eigenfunction(spectrum):
+    spectral_representation = np.eye(spectrum.total_num_lattice_points)[1, :]
+    real_space_representation = spectrum.transform(spectral_representation, "spectral", "real")
+    eigenfunction = spectrum.eigenfunction([1])
+    assert np.allclose(eigenfunction(*spectrum.lattice()), real_space_representation)
+
+
+def test_transform_with_correct_boundary_conditions(spectrum):
+    eigenfunctions = spectrum.eigenfunction(np.arange(spectrum.total_num_lattice_points).reshape(-1, 1))
```
Comment (Mayank447) [inline test/test_boundary_conditions.py:35]:
```suggestion
    eigenfunctions = spectrum.eigenfunction(np.arange(spectrum.total_num_lattice_points))
```

File: test/test_boundary_conditions.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+]
+
+
+@pytest.fixture(params=SPECTRA)
+def spectrum(request):
+    return request.param["type"](**request.param["config"])
+
+
+def test_eigenfunctions_obey_boundary_conditions(spectrum):
+    eigenfunctions = spectrum.eigenfunction(np.arange(spectrum.total_num_lattice_points).reshape(-1, 1))
```
Comment (Mayank447) [inline test/test_boundary_conditions.py:23]:
```suggestion
    eigenfunctions = spectrum.eigenfunction(np.arange(spectrum.total_num_lattice_points))
```

File: test/test_boundary_conditions.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    assert np.allclose(np.exp(2.0j * np.pi * spectrum.theta) * eigenfunctions(*spectrum.lattice()), eigenfunctions(spectrum.lattice()[0] + spectrum.L))
+
+
+def test_inverse_transform_results_in_eigenfunction(spectrum):
+    spectral_representation = np.eye(spectrum.total_num_lattice_points)[1, :]
```
Comment (Mayank447) [inline test/test_boundary_conditions.py:28]:
```suggestion
    spectral_representation = np.eye(spectrum.total_num_lattice_points)
```

File: test/test_boundary_conditions.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    real_space_representation = spectrum.transform(spectral_representation, "spectral", "real")
+    eigenfunction = spectrum.eigenfunction([1])
```
Comment (Mayank447) [inline test/test_boundary_conditions.py:30]:
```suggestion
    eigenfunction = spectrum.eigenfunction(np.arange(spectrum.total_num_lattice_points))
```

File: pseudospectral/spectra/derivative1D.py [hunk 6]
```diff
@@ -31,7 +33,8 @@ def eigenvalues(self):
         Private function to return the eigenvalues of the 1D derivative operator
         i.e. ik for the k-th eigenfunction exp(ikx) and k = 2*pi*m/L
         """
-        return I2PI * (np.fft.fftfreq(self.total_num_lattice_points, d=self.a))
+        tmp = I2PI * (np.fft.fftfreq(self.total_num_lattice_points, d=self.a) + self.theta / self.L)
+        return tmp
```
Comment (Mayank447) [inline pseudospectral/spectra/derivative1D.py:37]:
```suggestion
        return I2PI * (np.fft.fftfreq(self.total_num_lattice_points, d=self.a) + self.theta / self.L)
```

File: pseudospectral/spectra/derivative1D.py [hunk 7]
```diff
@@ -57,16 +58,15 @@ def transform(self, input_vector, input_basis, output_basis):
 
         # Perform the discrete Fast Fourier transform to go from real to spectral space
         elif input_basis == "real" and output_basis == "spectral":
-            return scipy.fft.fft(input_vector, norm="ortho") * np.sqrt(self.a)
+            return scipy.fft.fft(np.exp(-I2PI * self.theta / self.L * self.lattice("real")[0]) * input_vector, norm="ortho") * np.sqrt(self.a)
```
Comment (Mayank447) [inline pseudospectral/spectra/derivative1D.py:61]:
```suggestion
            premultiplier = np.exp(-I2PI * self.theta / self.L * self.lattice("real")[0])
            return scipy.fft.fft(premultiplier * input_vector, norm="ortho") * np.sqrt(self.a)
```

File: pseudospectral/spectra/derivative1D.py [hunk 8]
```diff
@@ -57,16 +58,15 @@ def transform(self, input_vector, input_basis, output_basis):
...
 
         # Perform the inverse discrete Fast Fourier transform to go from spectral to real space
         elif input_basis == "spectral" and output_basis == "real":
-            return scipy.fft.ifft(input_vector, norm="ortho") / np.sqrt(self.a)
+            return np.exp(I2PI * self.theta / self.L * self.lattice("real")[0]) * scipy.fft.ifft(input_vector, norm="ortho") / np.sqrt(self.a)
```
Comment (Mayank447) [inline pseudospectral/spectra/derivative1D.py:65]:
```suggestion
            postmultiplier = np.exp(I2PI * self.theta / self.L * self.lattice("real")[0])
            return postmultiplier * scipy.fft.ifft(input_vector, norm="ortho") / np.sqrt(self.a)
```

**Review by Mayank447 — CHANGES_REQUESTED** [review]
Few minor changes
